sandbox/cgroup/tracking.go

```diff
@@ -108,11 +114,26 @@ tryAgain:
 	//
 	// Verify the effective tracking cgroup and check that our scope name is
 	// contained therein.
-	path, err := cgroupProcessPathInTrackingCgroup(pid)
-	if err != nil {
-		return err
+	var pathAfter string
+	// XXX: It seems that systemd is racy and can not actually complete the
+	// request after returning from StartTransientUnit. Iterate for up to 500ms
+	// to see systemd perform the move. This is inherently racy.
+	// TODO: After pid-based tracking is removed, remove this spin logic as
+	// synchronicity is no longer relevant. It is only relevant during the
+	// transition period while both snap run and snap-confine manipulate
+	// cgroups.
+	for i := 0; i < 50; i++ {
+		pathAfter, err = cgroupProcessPathInTrackingCgroup(pid)
+		if err != nil {
+			return err
+		}
+		if pathBefore != pathAfter {
+			// Stop waiting as soon as the location in the cgroup hierarchy has changed.
+			break
```

Comment:
the break case doesn't seem to be tested

Reply:
Correct. I didn't test it because it was meant to be temporary. The next branch undoes this logic. Still, I thing each branch should stand on its own. I've added a test now.